libsel4vm/src/arch/arm/vgic/vgic.c

```diff
@@ -594,27 +582,32 @@ static int vgic_dist_disable_irq(struct vgic_dist_device *d, vm_vcpu_t *vcpu, in
 static int vgic_dist_set_pending_irq(struct vgic_dist_device *d, vm_vcpu_t *vcpu, int irq)
 {
     /* STATE c) */
-    struct gic_dist_map *gic_dist;
-    vgic_t *vgic;
-    struct virq_handle *virq_data;
+    struct gic_dist_map *gic_dist = vgic_priv_get_dist(d);
+    vgic_t *vgic = vgic_device_get_vgic(d);
 
-    gic_dist = vgic_priv_get_dist(d);
-    vgic = vgic_device_get_vgic(d);
+    struct virq_handle *virq_data = virq_find_irq_data(vgic, vcpu, irq);
 
-    virq_data = virq_find_irq_data(vgic, vcpu, irq);
-    /* If it is enabled, inject the IRQ */
-    if (virq_data && gic_dist->enable && is_enabled(gic_dist, irq, vcpu->vcpu_id)) {
-        int err;
-        DDIST("Pending set: Inject IRQ from pending set (%d)\n", irq);
+    if (!virq_data || !gic_dist->enable || !is_enabled(gic_dist, irq, vcpu->vcpu_id)) {
+        DDIST("IRQ not enabled (%d) on vcpu %d\n", irq, vcpu->vcpu_id);
+        return -1;
+    }
 
-        set_pending(gic_dist, virq_data->virq, true, vcpu->vcpu_id);
-        err = vgic_vcpu_inject_irq(d, vcpu, virq_data);
-        assert(!err);
+    if (is_pending(gic_dist, virq_data->virq, vcpu->vcpu_id)) {
+        return 0;
+    }
 
-        return err;
-    } else {
-        /* No further action */
-        DDIST("IRQ not enabled (%d) on vcpu %d\n", irq, vcpu->vcpu_id);
+    DDIST("Pending set: Inject IRQ from pending set (%d)\n", irq);
+    set_pending(gic_dist, virq_data->virq, true, vcpu->vcpu_id);
+
+    int err = vgic_vcpu_load_lr(d, vcpu, virq_data);
+    if (!err) {
+        return 0;
```

Comment:
The return code of `vgic_vcpu_load_lr()` is basically the one from `seL4_ARM_VCPU_InjectIRQ()`:
- seL4_DeleteFirst: this seems save to ignore, add to lr overflow queue
- seL4_IllegalOperation:should bubble up as a fatal error?
- seL4_InvalidCapability: this should bubble up as a fatal error?
- seL4_RangeError: this should bubble up as a fatal error?

Reply:
With the latest refactoring, all errors from seL4 kernel are considered fatal.